crowbar_framework/app/controllers/barclamp_controller.rb

```diff
@@ -271,103 +252,313 @@ def get_proposals_from_barclamps(barclamps)
-      Rails.logger.info "asking for proposal of: #{params.inspect}"
-      answer = @service_object.proposal_create params
-      Rails.logger.info "proposal is: #{answer.inspect}"
-      unless answer[0] == 200
-        flash[:alert] = answer[1]
+  add_help(:proposal_delete, [:id], [:delete])
+  def proposal_delete
+    code, message = @service_object.proposal_delete(
+      params[:id]
+    )
+
+    raise Crowbar::Error::NotFound if code == 404
+    respond_to do |format|
+      case code
+      when 200
+        format.html do
+          redirect_to(
+            barclamp_modules_path(
+              id: params[:controller]
+            )
+          )
+        end
+        format.json do
+          head :ok
+        end
+      else
+        format.html do
+          flash[:alert] = message
+
+          redirect_to(
+            barclamp_modules_path(
+              id: params[:controller]
+            )
+          )
+        end
+        format.json do
+          render json: { error: message }, status: code
+        end
       end
-    rescue StandardError => e
-      flash_and_log_exception(e)
     end
+  end
+
+  #
+  # Commit proposal
+  # Commit a specific proposal to apply it
+  # POST /crowbar/<barclamp-name>/<version>/proposals/commit/<barclamp-instance-name>
+  #
+  add_help(:proposal_commit, [:id], [:post])
+  def proposal_commit
+    code, message = @service_object.proposal_commit(
+      params[:id]
+    )
+
+    raise Crowbar::Error::NotFound if code == 404
     respond_to do |format|
-      format.html {
-        return redirect_to barclamp_modules_path id: params[:controller] if answer[0] != 200
-        redirect_to proposal_barclamp_path controller: controller, id: orig_id
-      }
-      format.xml  {
-        return render text: flash[:alert], status: answer[0] if answer[0] != 200
-        render xml: answer[1]
-      }
-      format.json {
-        return render text: flash[:alert], status: answer[0] if answer[0] != 200
-        render json: answer[1]
-      }
+      case code
+      when 200
+        format.html do
+          redirect_to(
+            proposal_barclamp_path(
+              controller: params[:controller],
+              id: params[:id]
+            )
+          )
+        end
+        format.json do
+          head :ok
+        end
+      when 202
+        format.html do
+          flash[:warning] = message
+
+          redirect_to(
+            proposal_barclamp_path(
+              controller: params[:controller],
+              id: params[:id]
+            )
+          )
+        end
+        format.json do
+          head :accepted
+        end
+      else
+        format.html do
+          flash[:alert] = message
+
+          redirect_to(
+            proposal_barclamp_path(
+              controller: params[:controller],
+              id: params[:id]
+            )
+          )
+        end
+        format.json do
+          render json: { error: message }, status: code
+        end
+      end
```

Comment:
I meant something like this for this block:
This makes it much more readable as the in the case statement only the differences occur, but no repetition.
```
  format.html do
    case code
    when 200
      headtag = :ok
    when 202
      flash[:warning] = message
      headtag = :accepted
    else
      flash[:alert] = message
      headtag = :error
    end

    redirect_to(
      proposal_barclamp_path(
        controller: params[:controller],
        id: params[:id]
      )
    )
  end

  format.json do
    head headtag unless headtag == :error
    render ( json: { error: message }, status: code ) if headtag == :error
  end
```